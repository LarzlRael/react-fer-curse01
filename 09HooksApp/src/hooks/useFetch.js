import { useEffect, useRef, useState } from "react"


export const useFetch = (url) => {

    const [state, setState] = useState({ data: null, loading: true, error: null });

    // using the useRef when the component is mounted
    const isMounted = useRef(true);
    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {

        setState({
            loading: true,
            error: null,
            data: null
        });

        fetch(url)
            .then(resp => resp.json())
            .then(data => {


                if (isMounted.current) {
                    // simulate 4 secods delay request
                    setState({
                        loading: false,
                        error: null,
                        data
                    })
                }
            }).catch(() => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la info prro',
                });
            } )
            
    }, [url])

    return state;
}
