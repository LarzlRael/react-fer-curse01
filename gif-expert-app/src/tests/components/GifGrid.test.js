import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';

jest.mock('../../hooks/useFetchGifs');

describe('Testing in <GifGrid /> component', () => {

    const category = 'One punch';

    test('should mostrar correctly', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });


        const wrapper = shallow(<GifGrid category={category} />)
        expect(wrapper).toMatchSnapshot();
    });

    test('should show items when the imagenes are loaded', () => {

        const gifs = [
            {
                id: 'abc',
                url: 'localhost',
                title: 'anything'
            },
            {
                id: 'cde',
                url: 'localhost:8080',
                title: 'anything here'
            },
        ];


        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });
        // supose there is information
        const wrapper = shallow(<GifGrid category={category} />)


        expect(wrapper.find('p').exists()).toBe(false);
        expect(wrapper.find('GifGridItem').length).toBe(gifs.length);


    });


})
