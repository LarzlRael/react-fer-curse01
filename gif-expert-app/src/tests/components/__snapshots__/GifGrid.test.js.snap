// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <GifGrid /> component should mostrar correctly 1`] = `
<Fragment>
  <h3
    className="card animate__animated animate__fadeIn"
  >
    One punch
  </h3>
  <p
    className="animate__animated animate__fadeIn"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
